src/constraint/4C_constraint_springdashpot.cpp

```diff
@@ -10,21 +10,228 @@
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& displacements,
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& displacement_offset,
+      const std::vector<double>& constant_offset,
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& velocity,
+      const std::vector<SpringDashpot>& spring_dashpot_evaluator, const double time_factor,
+      const Core::FE::GaussIntegration& gauss_integration, const std::vector<bool>& onoff,
+      std::optional<Core::LinAlg::Matrix<dim * Core::FE::num_nodes<celltype>, 1>>& residual_vector,
+      std::optional<Core::LinAlg::Matrix<dim * Core::FE::num_nodes<celltype>,
+          dim * Core::FE::num_nodes<celltype>>>& stiffness_matrix)
+  {
+    Core::Elements::for_each_gauss_point<celltype>(element_nodes, gauss_integration,
+        [&](const Core::LinAlg::Matrix<Core::FE::dim<celltype>, 1>& xi,
+            const Core::Elements::ShapeFunctionsAndDerivatives<celltype>& shape_functions,
+            const Core::Elements::JacobianMapping<celltype, dim>& jacobian_mapping,
+            double integration_factor, int gp)
+        {
+          Core::LinAlg::Matrix<dim, 1> displacement_at_gp;
+          displacement_at_gp.multiply_tn(displacements, shape_functions.values);
+
+          Core::LinAlg::Matrix<dim, 1> offset_at_gp;
+          offset_at_gp.multiply_tn(displacement_offset, shape_functions.values);
+
+          Core::LinAlg::Matrix<dim, 1> velocity_at_gp;
+          velocity_at_gp.multiply_tn(velocity, shape_functions.values);
+
+          for (unsigned axis_direction = 0; axis_direction < dim; ++axis_direction)
+          {
+            if (!onoff[axis_direction]) continue;
+
+            FOUR_C_ASSERT(spring_dashpot_evaluator.size() > axis_direction,
+                "Not enough spring-dashpot evaluators given!");
+
+            const auto& [force, force_deriv_disp, force_deriv_vel] =
+                spring_dashpot_evaluator[axis_direction](displacement_at_gp(axis_direction) -
+                                                             offset_at_gp(axis_direction) -
+                                                             constant_offset[axis_direction],
+                    velocity_at_gp(axis_direction));
+
+            // evaluate residual force
+            if (residual_vector)
+            {
+              for (int node = 0; node < Core::FE::num_nodes<celltype>; ++node)
+                residual_vector.value()(node * dim + axis_direction) +=
+                    shape_functions.values(node) * force * integration_factor;
+            }
+
+            // evaluate stiffness matrix
+            if (stiffness_matrix)
+            {
+              for (int i = 0; i < Core::FE::num_nodes<celltype>; ++i)
+              {
+                for (int j = 0; j < Core::FE::num_nodes<celltype>; ++j)
+                {
+                  stiffness_matrix.value()(i * dim + axis_direction, j * dim + axis_direction) +=
```

Comment:
Minor: Since you checked the optional for non-emptiness, this should just be
```suggestion
                  (*stiffness_matrix)(i * dim + axis_direction, j * dim + axis_direction) +=
```
so you don't pay for the empty check again.